docs/usage.rst

```diff
@@ -0,0 +1,259 @@
+
+Management Commands
+~~~~~~~~~~~~~~~~~~~
+
+One-off management commands that don't go through requests can
+also be instrumented in the ``manage.py`` file of a project
+using `pgactivity.contrib.execute_from_command_line`:
+
+.. code-block:: python
+
+    #!/usr/bin/env python
+    import os
+    import sys
+
+
+    if __name__ == "__main__":
+        os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")
+        from pgactivity.contrib import execute_from_command_line
+
+        execute_from_command_line(sys.argv)
+
+This function is a wrapper of Django's ``execute_from_command_line``
+and the ``command`` key added as context.
+It ignores ``runserver`` and ``runserver_plus`` by default.
+
+If it's not possible to use `pgactivity.contrib.execute_from_command_line`
+in your ``manage.py``, you can achieve the same functionality like so:
+
+.. code-block:: python
+
+    import contextlib
+    import sys
+    from django.core.management import execute_from_command_line
+
+
+    ignore_commands = ["runserver", "runserver_plus"]
+
+    if len(sys.argv) > 1 and not sys.argv[1] in ignore_commands:
+        activity_context = runtime.context(command=sys.argv[1])
+    else:
+        activity_context = contextlib.ExitStack()
+
+    with activity_context:
+        execute_from_command_line(sys.argv)
+
+Celery Tasks
+~~~~~~~~~~~~
+
+Celery tasks can also be instrumented like so:
+
+.. code-block:: python
+
+    import celery
+    import pgactivity
+
+    class Task(celery.Task):
+        def __call__(self, *args, **kwargs):
+            with pgactivity.context(task=self.name):
+                return super().__call__(*args, **kwargs)
+
+
+    # Override the celery task decorator for your application
+    app = create_celery_app('my-app')
+    task = app.task(base=Task)
+
+Setting Statement Timeouts
+--------------------------
+
+Using the `pgactivity.timeout` decorator and context manager to dynamically
+set Postgres's ``statement_timeout`` setting.
+
+For example, the following code ensures that no single query takes longer than
+500 milliseconds:
+
+.. code-block:: python
+
+    import pgactivity
+
+    with pgactivity.timeout(pgactivity.timedelta(milliseconds=500)):
+        # Issue queries. Any that exceed 500 milliseconds will raise an exception
+
+.. note::
+
+    `pgactivity.timedelta` is just a shortcut for Python's `datetime.timedelta`.
+
+The statement timeout will be applied locally to the connection and will
+not affect other queries. It can also be used as a decorator.
+
+.. warning::
+
+    `pgactivity.timeout` cannot be nested and will raise a `RuntimeError` if
+    it is nested. If running in a transaction and the transaction errors
+    before `pgactivity.timeout` exits, a warning will be printed and the
+    variable will not be flushed until the transaction is finished.
+    In other words, if a transaction save point is rolled back after
+    `pgactivity.timeout` exits, the variable will leak until the
```

Comment:
Yea here's some code to illustrate:

```
with transaction.atomic():
    with transaction.atomic():
        with pgactivity.timeout():
            # Do some stuff here to cause a DB exception, putting the transaction in an errored state

        # We exit pgactivity.timeout(). It can't flush the variable because the transaction is in
        # an errored state.

    # We continue here. The inner save point has been rolled back, but the LOCAL variable still
    # persists and leaks past the context manager
```

Reply:
It's even more difficult to nest it, especially when considering that you might have an outer call outside of a transaction and an inner call inside of a transaction. If the inner transaction is in an errored state, you can't reset the variable to its outer value.

Reply:
Ah yeah, nested atomics are just savepoints, I can see why even SET LOCAL would leak there. Might it make sense to just prevent someone using pgactivity.timeout in a _nested_ atomic?

Reply:
As mentioned offline, we can get around some of this stuff by using SQL hooks and prepending the variable to the statement like pghistory does. There are some drawbacks to the approach (like some SQL statements being outright ignored), but it works better for nesting and also has no performance penalties

Reply:
Following up one more time. It turns out that variables are rolled back if the save point is rolled back, making it possible to safely nest variables. Addressing this in a follow-up PR